Incorrent "Missing return statement" in presence of bad combination of decorated function call, definition order and NoReturn

Code to reproduce:

```python
from typing import Callable, NoReturn, TypeVar


def no_return() -> NoReturn:
    pass


def use_decorated_before_defined() -> int:
    decorated()
    if bool():
        return 0
    else:
        no_return()


T = TypeVar('T')


def wrapper(function: T) -> T:
    return function


def decorator() -> Callable[[T], T]:
    return wrapper


@decorator()
def decorated() -> None: pass


def use_decorated_after_defined() -> int:
    decorated()
    if bool():
        return 0
    else:
        no_return()


def use_wrapped_before_defined() -> int:
    wrapped()
    if bool():
        return 0
    else:
        no_return()


@wrapper
def wrapped() -> None: pass
```

I tested this both with mypy 0.750 and one from the current master branch (526a2187bbd2da35862899e7529394a74a6f6ced), running on CPython 3.7.5. The command line and the output:

```
% python -m mypy repro.py
repro.py:8: error: Missing return statement
Found 1 error in 1 file (checked 1 source file)
```

I'd expect no errors here. After some digging in mypy source code I believe `use_decorated_before_defined` can't be fully processed when it's type-checked (at least the first time it's entered) because of a) the call to `decorated` being looked at before `decorated` is defined and b) `decorated` being decorated (pardon my repetition) with a result of another function call which I suspect creates a difficult environment for mypy to deduce things and it defers further processing of `use_decorated_before_defined` (or at least part of it).

Then, when the call to `no_return` is encountered its type is detected as `Any` because of the processing being deferred per above, hence the error.

I left two similar variants of the function (`use_decorated_after_defined` and `use_wrapped_before_defined`) to help narrow things down as those don't trigger the error.

--- a/mypy/checker.py
+++ b/mypy/checker.py
@@ -1203,7 +1203,9 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
                 return_type = get_proper_type(return_type)
 
                 if self.options.warn_no_return:
-                    if not isinstance(return_type, (NoneType, AnyType)):
+                    if not self.current_node_deferred and not isinstance(
+                        return_type, (NoneType, AnyType)
+                    ):
                         # Control flow fell off the end of a function that was
                         # declared to return a non-None type and is not
                         # entirely pass/Ellipsis/raise NotImplementedError.
@@ -2431,6 +2433,7 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
         return (
             self.in_checked_function()
             and self.options.warn_unreachable
+            and not self.current_node_deferred
             and not self.binder.is_unreachable_warning_suppressed()
         )
 
@@ -4179,14 +4182,7 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
                                 # To support local variables, we make this a definition line,
                                 # causing assignment to set the variable's type.
                                 var.is_inferred_def = True
-                                # We also temporarily set current_node_deferred to False to
-                                # make sure the inference happens.
-                                # TODO: Use a better solution, e.g. a
-                                # separate Var for each except block.
-                                am_deferring = self.current_node_deferred
-                                self.current_node_deferred = False
                                 self.check_assignment(var, self.temp_node(t, var))
-                                self.current_node_deferred = am_deferring
                         self.accept(s.handlers[i])
                         var = s.vars[i]
                         if var:

--- a/test-data/unit/check-flags.test
+++ b/test-data/unit/check-flags.test
@@ -366,6 +366,22 @@ def f() -> NoReturn:  # E: Implicit return in function which does not return
   non_trivial_function = 1
 [builtins fixtures/dict.pyi]
 
+[case testNoReturnImplicitReturnCheckInDeferredNode]
+# flags: --warn-no-return
+from typing import NoReturn
+
+def exit() -> NoReturn: ...
+
+def force_forward_reference() -> int:
+    return 4
+
+def f() -> NoReturn:
+    x
+    exit()
+
+x = force_forward_reference()
+[builtins fixtures/exception.pyi]
+
 [case testNoReturnNoWarnNoReturn]
 # flags: --warn-no-return
 from mypy_extensions import NoReturn

--- a/test-data/unit/check-statements.test
+++ b/test-data/unit/check-statements.test
@@ -945,6 +945,18 @@ x = f()
 main:10: note: Revealed type is "builtins.int"
 main:15: note: Revealed type is "builtins.str"
 
+[case testExceptionVariableWithDisallowAnyExprInDeferredNode]
+# flags: --disallow-any-expr
+def f() -> int:
+    x
+    try:
+        pass
+    except Exception as ex:
+        pass
+    return 0
+x = f()
+[builtins fixtures/exception.pyi]
+
 [case testArbitraryExpressionAsExceptionType]
 import typing
 a = BaseException

--- a/test-data/unit/check-unreachable-code.test
+++ b/test-data/unit/check-unreachable-code.test
@@ -1397,3 +1397,20 @@ a or a  # E: Right operand of "or" is never evaluated
 1 and a and 1  # E: Right operand of "and" is never evaluated
 a and a  # E: Right operand of "and" is never evaluated
 [builtins fixtures/exception.pyi]
+
+[case testUnreachableFlagWithTerminalBranchInDeferredNode]
+# flags: --warn-unreachable
+from typing import NoReturn
+
+def assert_never(x: NoReturn) -> NoReturn: ...
+
+def force_forward_ref() -> int:
+    return 4
+
+def f(value: None) -> None:
+    x
+    if value is not None:
+        assert_never(value)
+
+x = force_forward_ref()
+[builtins fixtures/exception.pyi]

